be2-scala/src/main/scala/ch/epfl/pop/pubsub/PublishSubscribe.scala

```diff
@@ -5,30 +5,43 @@ import akka.actor.{ActorRef, ActorSystem}
 import akka.http.scaladsl.model.ws.{Message, TextMessage}
 import akka.pattern.AskableActorRef
 import akka.stream.FlowShape
-import akka.stream.scaladsl.{Flow, GraphDSL, Merge, Partition}
-import ch.epfl.pop.model.network.JsonRpcRequest
+import akka.stream.scaladsl.{Broadcast, Flow, GraphDSL, Merge, Partition}
+import ch.epfl.pop.decentralized.{ConnectionMediator, Monitor}
+import ch.epfl.pop.model.network.{JsonRpcRequest, JsonRpcResponse}
 import ch.epfl.pop.pubsub.graph._
-import ch.epfl.pop.pubsub.graph.handlers.{ParamsWithChannelHandler, ParamsWithMessageHandler}
+import ch.epfl.pop.pubsub.graph.handlers.{GetMessagesByIdResponseHandler, ParamsWithChannelHandler, ParamsWithMapHandler, ParamsWithMessageHandler}
 
 object PublishSubscribe {
 
   private var dbActorRef: AskableActorRef = _
 
   def getDbActorRef: AskableActorRef = dbActorRef
 
-  def buildGraph(mediatorActorRef: ActorRef, dbActorRefT: AskableActorRef, messageRegistry: MessageRegistry)(implicit system: ActorSystem): Flow[Message, Message, NotUsed] = Flow.fromGraph(GraphDSL.create() {
+  def buildGraph(
+      mediatorActorRef: ActorRef,
+      dbActorRefT: AskableActorRef,
+      messageRegistry: MessageRegistry,
+      monitorRef: ActorRef,
+      connectionMediatorRef: ActorRef,
+      isServer: Boolean
+  )(implicit system: ActorSystem): Flow[Message, Message, NotUsed] = Flow.fromGraph(GraphDSL.create() {
     implicit builder: GraphDSL.Builder[NotUsed] =>
       {
         import GraphDSL.Implicits._
 
-        val clientActorRef: ActorRef = system.actorOf(ClientActor.props(mediatorActorRef))
+        val clientActorRef: ActorRef = system.actorOf(ClientActor.props(mediatorActorRef, connectionMediatorRef, isServer))
         dbActorRef = dbActorRefT
 
+        if (isServer)
+          connectionMediatorRef ! ConnectionMediator.NewServerConnected(clientActorRef)
```

Comment:
I don't think that a message indicating that a new server connected should be emitted in a function called `buildGraph`.
It looks to me that either there should be a more meaningful place to call that portion of code, or build graph now starts to be doing a bit to many things at the same time (and would at least need a proper documentation indicating what it does and in which more global context it is supposed to be called).
Actually it is called during server's initialization and in `ConnectionMediator` when receiving a `ConnectTo` message, so that makes sense that it matches every time a new server actor is created, but we are getting quite far from the initial function intent of "building a DSL graph" so I would still root for a call somewhere else.

Reply:
You are right, this piece of code has followed me since week 4 or something I haven't thought about its sense since. Definitely not a great place. How about putting it in the ClientActor directly ? The clientActor is already the one telling ConnectionMediator when it leaves so it might as well tell it it arrived.

Reply:
taken care of in commit: 0ad00eea2447aa4c53d75c9556aee01db6306881

Reply:
Fixed !